crates/primitives/block/src/lib.rs

```diff
@@ -127,170 +81,388 @@ impl From<BlockId> for starknet_core::types::BlockId {
+    pub transactions: Vec<Transaction>,
+    pub receipts: Vec<TransactionReceipt>,
+    pub declared_classes: Vec<DeclaredClass>,
 }
 
-#[cfg(test)]
-mod tests {
-    use super::*;
+impl Validate for UnverifiedFullPendingBlock {
+    type Output = PreValidatedPendingBlock;
+    type ValidationError = BlockValidationError;
+    type Context = BlockValidationContext;
 
-    #[test]
-    fn test_maybe_pending_block_info() {
-        let tx_hashes_pending = vec![Felt::from(1), Felt::from(2)];
-        let pending = DeoxysPendingBlockInfo::new(PendingHeader::default(), tx_hashes_pending.clone());
-        let pending_as_maybe_pending: DeoxysMaybePendingBlockInfo = pending.clone().into();
-        let tx_hashes_not_pending = vec![Felt::from(3), Felt::from(4)];
-        let not_pending = DeoxysBlockInfo::new(Header::default(), tx_hashes_not_pending.clone(), Felt::from(5));
-        let not_pending_as_maybe_pending: DeoxysMaybePendingBlockInfo = not_pending.clone().into();
+    fn validate(mut self, context: &Self::Context) -> Result<Self::Output, Self::ValidationError> {
+        let classes = mem::take(&mut self.declared_classes);
+
+        let converted_classes = classes.convert()?;
+        let _tx_hashes = transaction_hashes(&self.receipts, &self.transactions, context, None)?;
 
-        assert_eq!(not_pending_as_maybe_pending.as_nonpending(), Some(&not_pending));
-        assert!(pending_as_maybe_pending.as_nonpending().is_none());
+        Ok(PreValidatedPendingBlock {
+            header: self.header,
+            transactions: self.transactions,
+            state_diff: self.state_diff,
+            receipts: self.receipts,
+            converted_classes,
+        })
+    }
+}
+
+/// Output of the [`crate::pre_validate`] step.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct PreValidatedBlock {
+    pub header: UnverifiedHeader,
+    pub transactions: Vec<Transaction>,
+    pub state_diff: StateDiff,
+    pub receipts: Vec<TransactionReceipt>,
+    pub commitments: ValidatedCommitments,
+    pub converted_classes: Vec<ConvertedClass>,
+
+    pub unverified_global_state_root: Option<Felt>,
+    pub unverified_block_hash: Option<Felt>,
+    pub unverified_block_number: Option<u64>,
+}
 
-        assert_eq!(pending_as_maybe_pending.as_pending(), Some(&pending));
-        assert!(not_pending_as_maybe_pending.as_pending().is_none());
+/// Output of the [`crate::pre_validate`] step.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct PreValidatedPendingBlock {
+    pub header: UnverifiedHeader,
+    pub transactions: Vec<Transaction>,
+    pub state_diff: StateDiff,
+    pub receipts: Vec<TransactionReceipt>,
+    pub converted_classes: Vec<ConvertedClass>,
+}
 
-        assert_eq!(pending_as_maybe_pending.as_block_id(), BlockId::Tag(BlockTag::Pending));
-        assert_eq!(not_pending_as_maybe_pending.as_block_id(), BlockId::Number(0));
+// ===================
+// TODO(akhercha) : move those *somewhere* else... not sure yet
```

Comment:
is that a todo you want fix before merge?

Reply:
If possible yup - that's related to the note at the bottom of my PR description, I could probably move them to the `transaction` primitive crate, but I like that they're close to where they're actually *only* used.

Kinda relates to the whole block primitives stuff I did - maybe it's just better in `block_import` like you did (close to where it's only used).

It really just depends on the future - if they'll be used elsewhere the primitives stuff make sense, else I think I'll revert that.
Wdyt?